nixos/earlyoom: fix desktop notification and add tests

oxalica opened the pull request:
As in title. Fixes #374959

Unresolved: The test produces a screenshot for manual assessment, but it is non-deterministic because the notification includes a PID. Should we keep or remove it?

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

oxalica committed:
nixosTests.earlyoom: migrate to runTest and set OOMScoreAdjust

oxalica committed:
nixosTests.earlyoom-notify: init for desktop notification

oxalica committed:
nixos/earlyoom: disable DynamicUser with notifications enabled

Otherwise D-Bus access will be blocked due to upstream issues.

- https://github.com/rfjakob/earlyoom/issues/270
- https://github.com/systemd/systemd/issues/22737

stephen-huan commented on nixos/modules/services/system/earlyoom.nix:
```diff
@@ -178,6 +178,11 @@ in
+      # Workaround: Upstream DynamicUser=true prevents notifications being sent to D-Bus.
+      # - earlyoom: https://github.com/NixOS/nixpkgs/issues/374959
+      # - systemd: https://github.com/systemd/systemd/issues/22737
+      serviceConfig.DynamicUser = mkDefault (!cfg.enableNotifications);
```
After reading https://github.com/systemd/systemd/issues/22737, I can confirm notifications work with broker enabled.

```suggestion
      serviceConfig.DynamicUser = mkDefault (
        !cfg.enableNotifications || config.services.dbus.implementation != "dbus"
      );
```

Whether you go with `config.services.dbus.implementation != "dbus"` or `config.services.dbus.implementation == "broker"` is whether you believe the bug is specific to dbus or a fix specific to broker. Currently there's only two possibilities for `config.services.dbus.implementation` so they're the same. I guess I would be optimistic and hope that future implementations won't have this problem. And you could add a test that it works with broker, at least. And maybe add a little note on this workaround in the description of `enableNotifications`?

oxalica replied:
Updated.

stephen-huan commented on nixos/tests/earlyoom-notify.nix:
```diff
@@ -0,0 +1,86 @@
+          output = machine.succeed('journalctl -u earlyoom.service -b0')
+          assert 'low memory! at or below SIGKILL limits' in output
+
+      with subtest("mako should get a notification"):
+        def hasOOMNotification(_last_try):
+          ret = json.loads(machine.succeed("${sudo} makoctl list"))
+          return any(
+            "Low memory! Killing process" in n["body"]["data"] \
+            for ns in ret["data"] \
+            for n in ns);
+        retry(hasOOMNotification)
+
+      machine.screenshot("got-oom-notification")
```
> Unresolved: The test produces a screenshot for manual assessment, but it is non-deterministic because the notification includes a PID. Should we keep or remove it?

If I understand correctly, the programmatic test is by string parsing `makoctl list` and `machine.screenshot("got-oom-notification")` is what produces `got-oom-notification.png` when running `nix build .#nixosTests.earlyoom-notify`. If the test works automatically and people can inspect the screenshot manually if they wish, why does it matter that the screenshot is non-deterministic if people are looking at it (and not computers)?

NixOS module tests are so futuristic !! Very cool.

stephen-huan reviewed (changes requested):
Thanks! This looks really good. Small nits about a dbus-broker workaround that allows keeping `DynamicUser = true` (which is marginally more secure, I guess).

oxalica committed:
{nixos,nixosTests}/earlyoom: enable DynamicUser for dbus-broker

The D-Bus issue only exists for default dbus implementation, so only
apply the workaround in that case.

stephen-huan reviewed (approved):
Thanks, looks good to me!
